File: trees/501. Find Mode in Binary Search Tree/findMode.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findMode(self, root):
        """
        Finds the modes using Inorder traversal benefiting from the structure of the BST:
        since duplicate values appear consecutively, using Inorder, we can track the current value,
        its frequency, and keep track of the maximum frequency

        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        stack = []
        current = root
        prev = None
        count = 0  # Frequency of the current node value
        max_count = 0  # Frequency of the current Mode
        result = []

        while current or stack:
            while current:
                stack.append(current)
                current = current.left
            
            current = stack.pop()

            # Count occurrences
            if prev and prev.val == current.val:
                count += 1
            else:
                count = 1  # Reset for new value

            # Store mode(s)
            if count > max_count:  # Found a new Mode
                max_count = count
                result = [current.val]  # Clear result and update it with the new Mode
            elif count == max_count:  # Found a different Mode with the same occurence
                result.append(current.val)
            
            prev = current
            current = current.right
        
        return result

File: trees/501. Find Mode in Binary Search Tree/findMode.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
    let frequency = {};
    let stack = [];
    let result = [];
    let current = root;
    let maxFreq = 0;

    while (current || stack.length > 0) {
        while (current) {
            stack.push(current);
            current = current.left;
        }

        current = stack.pop();

        if (current.val in frequency) frequency[current.val] += 1;
        else frequency[current.val] = 1;

        current = current.right;
    }

    for (let key in frequency) {
        if (frequency[key] > maxFreq) {
            maxFreq = frequency[key];
            result = [Number(key)];  // Keys in js obj are string
        } else if (frequency[key] === maxFreq) {
            result.push(Number(key));
        }
    }

    return result;
};